doc/source/development/rfc/rfc101_raster_dataset_threadsafety.rst

```diff
@@ -0,0 +1,474 @@
+
+- ``bool CanBeCloned(int nScopeFlags, bool bCanShareState) const``.
+  This method determines if a source dataset can be "cloned" (or re-opened).
+  It returns true for instances returned by GDALOpenEx, for instances of the MEM
+  driver if ``nScopeFlags`` == ``GDAL_OF_RASTER`` (and ``bCanShareState`` is
+  true for instances of the MEM driver)
+
+- ``std::unique_ptr<GDALDataset> Clone(int nScopeFlags, bool bCanShareState) const``.
+  This method returns a "clone" of the dataset on which it is called, and is used
+  by GDALThreadSafeDataset::ReferenceUnderlyingDataset() when a thread-local
+  dataset is needed.
+  Implementation of that method must be thread-safe.
+  The base implementation calls GDALOpenEx() reusing the dataset name, open flags
+  and open option. It is overloaded in the MEM driver to return a new instance
+  of MEMDataset, but sharing the memory buffers with the source dataset.
+
+No code in drivers, but the MEM driver, is modified by the candidate
+implementation.
+
+A few existing non-virtual methods of GDALDataset and GDALRasterBand have been
+made virtual (and overloaded by GDALProxyDataset and GDALProxyRasterBand),
+to avoid modifying state on the GDALThreadSafeRasterBand instance, which
+wouldn't be thread-safe.
+
+- :cpp:func:`GDALDataset::BlockBasedRasterIO`:
+  it interacts with the block cache
+- :cpp:func:`GDALRasterBand::GetLockedBlockRef`:
+  it interacts with the block cache
+- :cpp:func:`GDALRasterBand::TryGetLockedBlockRef`:
+  it interacts with the block cache
+- :cpp:func:`GDALRasterBand::FlushBlock`:
+  it interacts with the block cache
+- :cpp:func:`GDALRasterBand::InterpolateAtPoint`:
+  it uses a per-band cache
+- :cpp:func:`GDALRasterBand::EnablePixelTypeSignedByteWarning`: it should
+  already have been made virtual for GDALProxyRasterBand needs.
+
+Performance
+-----------
+
+The existing multireadtest utility that reads a dataset from multiple threads
+has been extended with a -thread_safe flag to asks to use GDAL_OF_THREAD_SAFE
+when opening the dataset in the main thread and use it in the worker threads,
+instead of the default behavior of opening explicitly a dataset in each thread.
+The thread-safe mode shows similar scalability as the default mode, sometimes
+with a slightly decreased efficiency, but not in a too problematic way.
+
+For example on a 20x20 raster:
+
+.. code-block:: shell
+
+    $ time multireadtest -t 4 -i 1000000 20x20.tif
+    real    0m2.084s
+    user    0m8.155s
+    sys     0m0.020s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 1000000 20x20.tif
+    real    0m2.387s
+    user    0m9.334s
+    sys     0m0.029s
+
+
+But on a 4096x4096 raster with a number of iterations reduced to 100. the
+thread_safe mode is slightly faster:
+
+.. code-block:: shell
+
+    $ time multireadtest -t 4 -i 100 4096x4096.tif
+    real    0m4.845s
+    user    0m18.666s
+    sys     0m0.084s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 100 4096x4096.tif
```

Comment:
I'm curious why this is faster, since it's doing pretty much the same thing.

Perhaps it's the effect of removing the `CPLSleep` call?

Reply:
> Perhaps it's the effect of removing the `CPLSleep` call?

The removal of the CPLSleep call applies to the "multireadtest -t 4" case too. But just retrying time measurement, I can see there's significant variation during runs, and sometimes classic is faster, sometimes pretty similar, so the above conclusion is not super relevant.

Reply:
chaned to "But on a 4096x4096 raster with a number of iterations reduced to 100, the timings between the default and thread_safe modes are very similar."